### PR TITLE
Feat add azure containers cache

### PR DESCRIPTION
**Pull request by mhaligowski**

### CODE REVIEW

File: internal/services/containers/container_registry_cache_rule_resource.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+func (ContainerRegistryCacheRule) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"name": {
+			Type:        pluginsdk.TypeString,
+			Required:    true,
+			Description: "The name of the cache rule.",
+		},
+
+		"registry": {
```
Comment (aristosvo) [inline internal/services/containers/container_registry_cache_rule_resource.go:34]: I would opt for `container_registry_id` here, which contains both the resource group name and registry name, as input.

How to work with that ID is shown here: https://github.com/hashicorp/terraform-provider-azurerm/blob/46a95c1b799aca661d980992e65779e8576178e7/internal/services/containers/container_registry_task_resource.go#L126

File: internal/services/containers/container_registry_cache_rule_resource.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+				return fmt.Errorf("deleting Container Registry Cache Rule %s: %+v", *id, err)
+			}
+
+			return nil
+		},
+	}
+}
+
+func (ContainerRegistryCacheRule) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return validate.ContainerRegistryCacheRuleID
```
Comment (aristosvo) [inline internal/services/containers/container_registry_cache_rule_resource.go:215]: I believe you can replace the `validate` package with the content of `cacherules` in the new SDK, like this:
```suggestion
	return cacherules.ValidateCacheRuleID
```

File: internal/services/containers/resourceids.go [hunk 3]
```diff
@@ -5,5 +5,6 @@ package containers
 
 //go:generate go run ../../tools/generator-resource-id/main.go -path=./ -name=Cluster -id=/subscriptions/12345678-1234-9876-4563-123456789012/resourceGroups/resGroup1/providers/Microsoft.ContainerService/managedClusters/cluster1
 //go:generate go run ../../tools/generator-resource-id/main.go -path=./ -name=NodePool -id=/subscriptions/12345678-1234-9876-4563-123456789012/resourceGroups/resGroup1/providers/Microsoft.ContainerService/managedClusters/cluster1/agentPools/pool1
+//go:generate go run ../../tools/generator-resource-id/main.go -path=./ -name=ContainerRegistryCacheRule -id=/subscriptions/12345678-1234-9876-4563-123456789012/resourceGroups/group1/providers/Microsoft.ContainerRegistry/registries/registry1/cacheRules/rule1
```
Comment (aristosvo) [inline internal/services/containers/resourceids.go:8]: If you don't need the `validate` package for ID validation, you can remove this resource ID entry as well!

File: internal/services/containers/client/client.go [hunk 4]
```diff
@@ -37,6 +40,12 @@ type Client struct {
 }
 
 func NewContainersClient(o *common.ClientOptions) (*Client, error) {
+	cacheRulesClient, err := cacherules.NewCacheRulesClientWithBaseURI(o.Environment.ResourceManager)
```
Comment (aristosvo) [inline internal/services/containers/client/client.go:43]: I don't think this declaration of the cacheRulesClient is necessary for the new SDK clients anymore

File: internal/services/containers/container_registry_cache_rule_resource_test.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	"github.com/hashicorp/terraform-provider-azurerm/internal/acceptance/check"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/clients"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/containers/validate"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/utils"
+)
+
+type ContainerRegistryCacheRuleResource struct{}
+
+func TestAccContainerRegistryCacheRuleName_validation(t *testing.T) {
```
Comment (aristosvo) [inline internal/services/containers/container_registry_cache_rule_resource_test.go:19]: This `validate` function is already tested in its own package (and you won't need it after all)

**Review by aristosvo — CHANGES_REQUESTED** [review]
Hi @mhaligowski! Thanks for all this! I gave you a few pointers to start with, as I don't expect a review from the HC maintainers these days, they will have probably a few more remarks.

There are two tests failing as well:
```
------- Stdout: -------
=== RUN   TestAccContainerRegistryCacheRule_basic
=== PAUSE TestAccContainerRegistryCacheRule_basic
=== CONT  TestAccContainerRegistryCacheRule_basic
    testcase.go:113: Step 1/2 error: Error running pre-apply refresh: exit status 1
        Error: Unsupported argument
          on terraform_plugin_test.tf line 35, in resource "azurerm_container_registry_cache_rule" "test":
          35:   resource_group_name = azurerm_resource_group.test.name
        An argument named "resource_group_name" is not expected here.
--- FAIL: TestAccContainerRegistryCacheRule_basic (2.78s)
FAIL
```
```
------- Stdout: -------
=== RUN   TestAccContainerRegistryCacheRule_requiresImport
=== PAUSE TestAccContainerRegistryCacheRule_requiresImport
=== CONT  TestAccContainerRegistryCacheRule_requiresImport
    testcase.go:113: Step 1/2 error: Error running pre-apply refresh: exit status 1
        Error: Unsupported argument
          on terraform_plugin_test.tf line 35, in resource "azurerm_container_registry_cache_rule" "test":
          35:   resource_group_name = azurerm_resource_group.test.name
        An argument named "resource_group_name" is not expected here.
--- FAIL: TestAccContainerRegistryCacheRule_requiresImport (3.10s)
FAIL
```

Hope you enjoy the last bits of 2023!

**Comment by mhaligowski** [comment]
> Hi @mhaligowski! Thanks for all this! I gave you a few pointers to start with, as I don't expect a review from the HC maintainers these days, they will have probably a few more remarks.
> 
> There are two tests failing as well:
> 
> ```
> ------- Stdout: -------
> === RUN   TestAccContainerRegistryCacheRule_basic
> === PAUSE TestAccContainerRegistryCacheRule_basic
> === CONT  TestAccContainerRegistryCacheRule_basic
>     testcase.go:113: Step 1/2 error: Error running pre-apply refresh: exit status 1
>         Error: Unsupported argument
>           on terraform_plugin_test.tf line 35, in resource "azurerm_container_registry_cache_rule" "test":
>           35:   resource_group_name = azurerm_resource_group.test.name
>         An argument named "resource_group_name" is not expected here.
> --- FAIL: TestAccContainerRegistryCacheRule_basic (2.78s)
> FAIL
> ```
> 
> ```
> ------- Stdout: -------
> === RUN   TestAccContainerRegistryCacheRule_requiresImport
> === PAUSE TestAccContainerRegistryCacheRule_requiresImport
> === CONT  TestAccContainerRegistryCacheRule_requiresImport
>     testcase.go:113: Step 1/2 error: Error running pre-apply refresh: exit status 1
>         Error: Unsupported argument
>           on terraform_plugin_test.tf line 35, in resource "azurerm_container_registry_cache_rule" "test":
>           35:   resource_group_name = azurerm_resource_group.test.name
>         An argument named "resource_group_name" is not expected here.
> --- FAIL: TestAccContainerRegistryCacheRule_requiresImport (3.10s)
> FAIL
> ```
> 
> Hope you enjoy the last bits of 2023!

I did, thanks :) I'll try pushing it through in the first week of 2024 :)

**Comment by katbyte** [comment]
@mhaligowski - is this PR now ready to be reviewed? it has been in draft for a while

**Comment by stephybun** [comment]
Going to close this for the moment since we haven't heard back. Let us know when you're ready to come back to this @mhaligowski and we can re-open this!

**Pull request closed without merging**